AutomaticQC/imosEchoRangeSetQC.m

```diff
@@ -0,0 +1,167 @@
+    if strncmpi(paramName, 'VCUR', 4),  idVcur = i; end
+    if strcmpi(paramName, 'WCUR'),      idWcur = i; end
+    if strcmpi(paramName, 'CSPD'),      idCspd = i; end
+    if strncmpi(paramName, 'CDIR', 4),  idCdir = i; end
+    for j=1:4
+        cc = int2str(j);
+        if strcmpi(paramName, ['ABSIC' cc]), idABSIC{j} = i; end
+    end
+end
+
+% check if the data is compatible with the QC algorithm
+idMandatory = (idUcur | idVcur | idWcur | idCspd | idCdir);
+for j=1:4
+    idMandatory = idMandatory & idABSIC{j};
+end
+if ~idMandatory, return; end
+
+% let's get the associated vertical dimension
+idVertDim = sample_data.variables{idABSIC{1}}.dimensions(2);
+if strcmpi(sample_data.dimensions{idVertDim}.name, 'DIST_ALONG_BEAMS')
+    disp(['Warning : imosEchoRangeSetQC applied with a non tilt-corrected ABSICn (no bin mapping) on dataset ' sample_data.toolbox_input_file]);
+end
+
+qcSet           = str2double(readProperty('toolbox.qc_set'));
+badFlag         = imosQCFlag('bad',             qcSet, 'flag');
+goodFlag        = imosQCFlag('good',            qcSet, 'flag');
+rawFlag         = imosQCFlag('raw',             qcSet, 'flag');
+
+%Pull out echo intensity
+sizeData = size(sample_data.variables{idABSIC{1}}.data);
+ea = nan(4, sizeData(1), sizeData(2));
+for j=1:4
+    ea(j, :, :) = sample_data.variables{idABSIC{j}}.data;
+end
+
+% read in filter parameters
+propFile  = fullfile('AutomaticQC', 'imosEchoRangeSetQC.txt');
+ea_fishthresh = str2double(readProperty('ea_fishthresh',   propFile));
+
+% read dataset QC parameters if exist and override previous 
+% parameters file
+currentQCtest = mfilename;
+ea_fishthresh = readDatasetParameter(sample_data.toolbox_input_file, currentQCtest, 'ea_fishthresh', ea_fishthresh);
+
+paramsLog = ['ea_fishthresh=' num2str(ea_fishthresh)];
+
+
+% Run QC
+% Following code is adapted from the UWA 'adcpfishdetection.m' code
+
+[n, t, m]=size(ea); % m depth cells, n (4) beams, t timesteps
+% same flags are given to any variable
+bad_ea = ones(n,t,m,'int8')*rawFlag;
+
+% matrix operation of the UW original code which loops over each timestep
+[B, Ix]=sort(ea,1); %sort echo from highest to lowest along each bin
+    %step one - really only useful for data in beam coordinates. If one
+    %beam fails, then can do 3-beam solutions
+if unique(sample_data.meta.fixedLeader.coordinateTransform) ~=7 %7 is the value for ENU setup. Not sure what it is for beam coords. Can be refined here.
+    df = B(4,:,:)-B(1,:,:); %get the difference from highest value to lowest
+    ind=df>ea_fishthresh; % problematic depth cells
+    bad_ea(ind) = true; %flag these as bad (individual cells)
+end
+%step 2: useful for data in both beam and ENU coordinates. Flags entire
+%bin of velocity data
+df = B(4,:,:)-B(2,:,:); %get the difference from highest value to second lowest
+ind=df>ea_fishthresh; % problematic depth cells
+bad_ea(:,ind) = true; %flag the entire bin (all beams) as bad
+
```

Comment:
we could improve speed/memory here by just using "max/min" instead of sort.
```
maxecho = max(ea,[],1);
minecho = min(ea,[],1);
invalid = (maxecho-minecho) > ea_fishtresh;
...
```
while for the ENU we could just use mink function:

```
secondmin = mink(ea,2,1)(2);
...
```